### PR TITLE
Updating PennyLane to current version

### PR DESCRIPTION
**Pull request by bdoolittle**
This pull request implements breaking changes to resolve incompatibility with v0.23+ of PennyLane. The solution is to not store network settings in a ragged array, but to store them in a 1D array instead. As a result the construction of qnode settings must be altered.

## Breaking Changes

* Optimization settings are no longer stored in a ragged list structure. For compatibility with current versions of PennyLane, we place all network settings into a 1D list.
* The `NetworkAnsatz` class create randomized or zero settings that are in a 1D list.
* Static settings are no longer set at the node-level, but instead specified when creating a settings list using `qnetvo.NetworkAnsatz.rand_network_settings()`.
* cost function constructors are renamed with a `_fn` postscript to indicate a function is constructed.

### CODE REVIEW

File: demos/magic_squares_game_optimization.ipynb [hunk 1]
```diff
@@ -99,21 +99,13 @@
...
     {
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "opt score :  0.8888798713795558\n",
-      "CPU times: user 30.6 s, sys: 102 ms, total: 30.7 s\n",
-      "Wall time: 30.7 s\n"
+      "opt score :  0.8888798713795557\n",
+      "CPU times: user 42.5 s, sys: 183 ms, total: 42.6 s\n",
+      "Wall time: 42.9 s\n"
```
Comment (trbromley) [inline demos/magic_squares_game_optimization.ipynb:108]: Wonder if this is indicative of a slowdown, or maybe you were running on a different machine/laptop unplugged?

File: docs/requirements.txt [hunk 2]
```diff
@@ -4,4 +4,4 @@ Pygments>=2.7.4
 m2r2==0.3.2
 sphinxcontrib-email==0.3.5
 furo
-pennylane==0.22
+pennylane==0.26
```
Comment (trbromley) [inline docs/requirements.txt:7]: Should there be a period at the top of this file?

File: src/qnetvo/network_ansatz.py [hunk 3]
```diff
@@ -40,18 +40,14 @@ class PrepareNode(NoiseNode):
     :param num_settings: The number of settings that the quantum function accepts.
     :type num_settings: int
 
-    :param static_settings: A set of fixed settings for the preparation node.
-    :type static_settings: optional, array-like
-
```
Comment (trbromley) [inline src/qnetvo/network_ansatz.py:45]: Nice that we can get rid of static settings, and just specify directly with `requires_grad`!

File: src/qnetvo/network_ansatz.py [hunk 4]
```diff
@@ -191,6 +189,31 @@ def ansatz_circuit(settings):
 
         return ansatz_circuit
 
+    def get_network_parameter_partitions(self):
+        """
+        A ragged array containing tuples that specify how to partition a 1D array
```
Comment (trbromley) [inline src/qnetvo/network_ansatz.py:194]:
```suggestion
        A nested list containing tuples that specify how to partition a 1D array
```
Very minor, but I found "ragged array" confusing.

File: src/qnetvo/network_ansatz.py [hunk 5]
```diff
@@ -283,64 +319,56 @@ def circuit(settings):
...
+        If fixed settings are specified, then they are marked as ``requires_grad=False`` and
+        not differentatiated during optimzation.
 
-        :returns: A nested list containing settings for each network node.
-                  ``PreparNode`` settings are listed under index ``0`` while
-                  ``MeasureNode`` settings are listed under index ``1``.
-                  The measure and prepare layers settings are a list of numpy arrays
-                  where the dimension of each array is ``(num_inputs, num_settings)``.
-        :rtype: list[list[np.array]]
+        :param fixed_setting_ids: The ids of settings that are held constant during optimization.
```
Comment (trbromley) [inline src/qnetvo/network_ansatz.py:327]:
```suggestion
        :param fixed_setting_ids: The ids of settings that are held constant during optimization.
        Also requires `fixed_settings` to be provided.
```

File: src/qnetvo/network_ansatz.py [hunk 6]
```diff
@@ -283,64 +319,56 @@ def circuit(settings):
...
+        return rand_settings
 
-    def tf_rand_scenario_settings(self):
+    def tf_rand_network_settings(self, fixed_setting_ids=[], fixed_settings=[]):
         """Creates a randomized settings array for the network ansatz using TensorFlow
         tensor types.
 
-        :returns: See :meth:`qnetvo.NetworkAnsatz.rand_scenario_settings` for details.
-        :rtype: list[list[tf.Tensor]]
+        :param fixed_setting_ids: The ids of settings that are held constant during optimization.
```
Comment (trbromley) [inline src/qnetvo/network_ansatz.py:350]:
```suggestion
        :param fixed_setting_ids: The ids of settings that are held constant during optimization.
        Also requires `fixed_settings` to be provided.
```

File: src/qnetvo/network_ansatz.py [hunk 7]
```diff
@@ -283,64 +319,56 @@ def circuit(settings):
...
+        :type fixed_setting_ids: *optional* List[Int]
+
+        :param fixed_settings: The constant values for fixed settings.
+        :type fixed_settings: *optional* List[Float]
+
+        :returns: See :meth:`qnetvo.NetworkAnsatz.rand_network_settings` for details.
```
Comment (trbromley) [inline src/qnetvo/network_ansatz.py:356]:
```suggestion
        :returns: A 1D list of ``tf.Variable`` and ``tf.constant`` scalar values.
```

File: src/qnetvo/network_ansatz.py [hunk 8]
```diff
@@ -283,64 +319,56 @@ def circuit(settings):
...
-            for node in self.prepare_nodes
-        ]
-        measure_settings = [
-            np.zeros(node.settings_dims) if len(node.static_settings) == 0 else np.array([[]])
-            for node in self.measure_nodes
-        ]
-
-        return [prepare_settings, measure_settings]
+        num_settings = self.parameter_partitions[-1][-1][-1][-1]
+        return [np.array(0) for _ in range(num_settings)]
```
Comment (trbromley) [inline src/qnetvo/network_ansatz.py:374]:
```suggestion
        return [np.array(0, requires_grad=True) for _ in range(num_settings)]
```
Maybe better to be explicit. I recall discussing the merits of having `requires_grad=False` by default in PL.

File: test/chsh_gradient_descent_test.py [hunk 9]
```diff
@@ -89,18 +116,20 @@ def test_parallel_chsh_natural_gradient_descent(self):
         ]
 
         chsh_ansatz = qnet.NetworkAnsatz(prepare_nodes, measure_nodes)
-        chsh_cost = qnet.chsh_inequality_cost(chsh_ansatz)
-        chsh_grad_fn = qnet.chsh_natural_grad(chsh_ansatz)
+        chsh_cost = qnet.chsh_inequality_cost_fn(chsh_ansatz)
+        # chsh_grad_fn = qnet.chsh_natural_grad(chsh_ansatz)
```
Comment (trbromley) [inline test/chsh_gradient_descent_test.py:120]:
```suggestion
```

File: test/cost/linear_inequalities_test.py [hunk 10]
```diff
@@ -12,51 +12,43 @@ def example_ansatz(self):
             qnet.PrepareNode(4, [1, 2], qnet.local_RY, 2),
         ]
         meas_nodes = [
-            qnet.MeasureNode(1, 2, [0, 1, 2], qnet.local_RY, 3),
+            # qnet.MeasureNode(1, 2, [0, 1, 2], qnet.local_RY, 3),
```
Comment (trbromley) [inline test/cost/linear_inequalities_test.py:15]:
```suggestion
```

File: test/cost/nlocal_star_bell_inequality_test.py [hunk 11]
```diff
@@ -69,36 +61,37 @@ def test_trilocal_star_cost(self, parallel_flag, nthreads):
...
 
-        assert np.isclose(trilocal_22_cost(ideal_settings), -np.sqrt(2))
+        assert np.isclose(trilocal_22_cost(*ideal_settings), -np.sqrt(2))
 
     @pytest.mark.parametrize("parallel_flag, nthreads", [(False, 4), (True, 4), (True, 5)])
     def test_bilocal_star_22_cost_gradient_descent(self, parallel_flag, nthreads):
         bilocal_star_ansatz = self.bilocal_star_ry_ansatz()
 
-        np.random.seed(45)
+        # np.random.seed(45)
```
Comment (trbromley) [inline test/cost/nlocal_star_bell_inequality_test.py:89]: Do we want to keep the seed here?

**Review by trbromley — APPROVED** [review]
Thanks @bdoolittle! I've left a few comments, but otherwise great to get QNetVO updated to use the latest PL version :rocket:

File: docs/requirements.txt (hunk 2 above)
Comment (bdoolittle) [reply docs/requirements.txt:7]: shouldn't cause a problem. I removed the line. We'll see if the docs build properly in main

File: demos/magic_squares_game_optimization.ipynb (hunk 1 above)
Comment (bdoolittle) [reply demos/magic_squares_game_optimization.ipynb:108]: Unfortunately, I was able to reproduce this slow down on a number of examples. The gradient descent optimization takes roughly 33% longer in v0.2 than it did in the previous update using pennylane 0.23.

My guess is that the performance problem might be caused by storing the settings as individual tensors withe requires grad set independently for each element. Otherwise, has pennylane experienced a general slowdown over the past few releases?

**Comment by bdoolittle** [comment]
@trbromley I made some edits based on your comments.

Some items of note are:
* The dot at the top of the requirements.txt file is needed for the CI doc checks to pass, but not local integration.
* I was able to reproduce the slowdown you caught. Gradient descent optimization Is about 33% slower in qnetvo v0.2 than v0.1.4. I'm not too worried about this right now, but we might have to modify how we handle fixed settings in the future.

**Review by trbromley — APPROVED** [review]
Thanks @bdoolittle!
